Core/GDCore/Project/EffectsContainer.h

```diff
@@ -0,0 +1,99 @@
+/*
+ * GDevelop Core
+ * Copyright 2008-2016 Florian Rival (Florian.Rival@gmail.com). All rights
+ * reserved. This project is released under the MIT License.
+ */
+#ifndef GDCORE_EFFECTS_CONTAINER_H
+#define GDCORE_EFFECTS_CONTAINER_H
+#include <memory>
+#include <vector>
+
+#include "GDCore/String.h"
+
+namespace gd {
+class Effect;
+}  // namespace gd
+
+namespace gd {
+
+/**
+ * \brief Contains the effect applied to other entities (ie, Layer or Object)
+ *
+ * \see gd::EffectsContainer
+ * \ingroup PlatformDefinition
+ */
+class GD_CORE_API EffectsContainer {
```

Comment:
So this is indeed the idea, though I'm surprised because there seems to be a missing thing in the existing gd::Layer: a custom copy constructor and assignement operator.

The reason is that by default, the compiler will generate a default copy constructor (used when you do `gd::EffectsContainer newCopy = otherContainer;`) or assignement operator (used when you do `someContainer = otherContainer;`).

By default, C++ generates a constructor/operator that just iterate on the members, and copy them. That's fair enough! And almost always what we want
BUT for a vector of shared_ptr, it will copy all shared_ptr... which means that if you do `someContainer = otherContainer;`, then `someContainer` and `otherContainer` will now have two different vector in memory (so far so good), containing different pointers (so far so good), but all these pointers will have the same values! So they will point to the same effects!
(Imagine the confusion: you duplicate a layer, then you edit some effects, and the effects of the other layers are changed too!).

What a pain. This is because the "semantics" of copying a shared_ptr is "just create another pointer to the SAME object" (much like actually in JavaScript :)). So what do we do? We re-implement these copy constructor/assignement, so that they have a semantic which is to COPY the objects of the container.

Usually I do something like this (search VariablesContainer.cpp):
```
VariablesContainer::VariablesContainer(const VariablesContainer& other) {
  Init(other);
}

VariablesContainer& VariablesContainer::operator=(
    const VariablesContainer& other) {
  if (this != &other) Init(other);

  return *this;
}

void VariablesContainer::Init(const gd::VariablesContainer& other) {
  variables.clear();
  for (auto& it : other.variables) {
    variables.push_back(
        // In your case you won't need the std::make_pair, because we're not using a pair, just a shared_ptr
        std::make_pair(it.first, std::make_shared<gd::Variable>(*it.second)));
  }
}
```

With this, we should be safe. Copying a container will work "as we could expect" by copying everything inside.

Reply:
> So they will point to the same effects!
(Imagine the confusion: you duplicate a layer, then you edit some effects, and the effects of the other layers are changed too!).

I've been dealing with the exact thing for the last 10 mins 😆. Apart from that, I've added serialization and everything seems to be working well. I'll push the code after making changes these changes.

Reply:
> I've been dealing with the exact thing for the last 10 mins 😆. 

I'm almost glad you did ;) You have to get burned at least once by this kind of error to really understand and remember it 😬 But be happy, you've now passed a huge milestone in your C++/shared pointers knowledge! :D